### PR TITLE
Invoke workflow

### LINKED ISSUE #351: Update mechanism to get status of workflows from Galaxy
## Describe your idea for a new feature

Currently, we make use of an old API method in Galaxy which runs workflows and does not return until all jobs in Galaxy are scheduled. This method in the Galaxy API is deprecated in favour of a method which returns right away before all jobs are scheduled (`run` vs. `invoke` - https://bioblend.readthedocs.io/en/latest/api_docs/galaxy/all.html#bioblend.galaxy.workflows.WorkflowClient.run_workflow).  Updating to use `invoke` will require changes to how we check for the status of running workflows in IRIDA.

## Describe how you would access this feature?

This would involve updating the [blend4j](https://github.com/galaxyproject/blend4j) project, and then updating the IRIDA/Galaxy interaction code. This would particularly impact the IRIDA pipeline status display:

image

## Additional information

Related to issue #328

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
## Description of changes
Update deprecated runWorkflow to invokeWorkflow blend4j calls.
Switched to packaged jar instead of pulling blend4j from maven repo

## Related issue
Fixes #351 

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [x] Tests added (or description of how to test) for any new features.
* [ ] User documentation updated for UI or technical changes.

### CODE REVIEW

**Comment by apetkau** [comment]
So, there's another issue that we will need to address that only comes up every so often (so it makes it hard to catch).

# Problem

Basically, the new API call to schedule Galaxy workflows `invokeWorkflow()` does not block while it waits for all jobs to be scheduled in Galaxy. The older API call `runWorkflow()` did block until all jobs where scheduled in Galaxy.

This has the consequence that sometimes, depending on how long it takes for Galaxy to schedule jobs, it's possible for IRIDA to mistakenly think a pipeline is finished when it really just hasn't had enough time to schedule all jobs in Galaxy. This would appear in IRIDA as a failed pipeline (status ERROR). You would also likely notice an exception like the following:

```
15 May 2020 18:36:45,787 ERROR ca.corefacility.bioinformatics.irida.service.analysis.execution.AnalysisExecutionServiceAspect:65 - Error occured for submission: AnalysisSubmission [id=115, name=Automated SISTRTyping - SH14-023, submitter=admin, workflowId=92ecf046-ee09-4271-b849-7a82625d6b60, analysisState=COMPLETING, analysisCleanedState=NOT_CLEANED] changing to state ERROR
ca.corefacility.bioinformatics.irida.exceptions.ExecutionManagerDownloadException: Could not download dataset identified by historyId=241d84796a24640a, datasetId=c7b8ef07ab33d346 to destination=/tmp/analysis-output952968070229116930/sistr-cgmlst-profiles.csv
        at ca.corefacility.bioinformatics.irida.pipeline.upload.galaxy.GalaxyHistoriesService.downloadDatasetTo(GalaxyHistoriesService.java:340)
        at ca.corefacility.bioinformatics.irida.service.analysis.workspace.galaxy.AnalysisWorkspaceServiceGalaxy.buildOutputFile(AnalysisWorkspaceServiceGalaxy.java:140)
        at ca.corefacility.bioinformatics.irida.service.analysis.workspace.galaxy.AnalysisWorkspaceServiceGalaxy.getAnalysisResults(AnalysisWorkspaceServiceGalaxy.java:381)
        at ca.corefacility.bioinformatics.irida.service.analysis.execution.galaxy.AnalysisExecutionServiceGalaxyAsync.transferAnalysisResults(AnalysisExecutionServiceGalaxyAsync.java:191)
...
Caused by: com.sun.jersey.api.client.UniformInterfaceException: GET http://localhost:48888/api/histories/241d84796a24640a/contents/c7b8ef07ab33d346/display?key=admin&to_ext=csv returned a response status of 500 Internal Server Error
```

Essentially, it's trying to get one of the output files from Galaxy, but the file doesn't exist.

If we look at the database, we can see:

```
select modified_date,analysis_state from analysis_submission_AUD where id = 115;
+---------------------+------------------+
| modified_date       | analysis_state   |
+---------------------+------------------+
| 2020-05-15 16:34:57 | NEW              |
| 2020-05-15 16:34:57 | NEW              |
| 2020-05-15 16:34:57 | NEW              |
| 2020-05-15 18:36:14 | PREPARING        |
| 2020-05-15 18:36:14 | PREPARED         |
| 2020-05-15 18:36:23 | SUBMITTING       |
| 2020-05-15 18:36:35 | RUNNING          |
| 2020-05-15 18:36:36 | FINISHED_RUNNING |
| 2020-05-15 18:36:45 | COMPLETING       |
| 2020-05-15 18:36:45 | ERROR            |
+---------------------+------------------+
```

If you notice the time from `RUNNING` to `FINISHED_RUNNING`, it only differs by 1 second. Hence, this pipeline was put into the `FINISHED_RUNNING` state almost immediately after it switched to `RUNNING`. When I looked into Galaxy on some of these pipelines I noticed they were still running in Galaxy (even though IRIDA thought they were finished).

# Cause

The code we were using to determine if a pipeline is completed in IRIDA is here https://github.com/phac-nml/irida/blob/6af8e5465830a4b78ac856a0fe2db4472190b996/src/main/java/ca/corefacility/bioinformatics/irida/service/impl/AnalysisExecutionScheduledTaskImpl.java#L274

This ultimately calls this code https://github.com/phac-nml/irida/blob/6af8e5465830a4b78ac856a0fe2db4472190b996/src/main/java/ca/corefacility/bioinformatics/irida/service/impl/AnalysisExecutionScheduledTaskImpl.java#L20

which checks the status of the History in Galaxy for a state of `OK`. Now, it's possible to get a state of `OK` any time there are no queued or running jobs in a history (either because all jobs are finished, or no jobs have been scheduled yet). In this case, the error is caused by the reporting of a status of `OK` for the history because the workflow has not finished scheduling in Galaxy (which is mistakenly interpreted by IRIDA as the workflow having completed successfully).

This never occurred in our previous code because `runWorkflow()` would never return until all jobs are scheduled from the workflow (and so all appear in the History).

# Solution

The solution will require us to modify the code to make sure that we only switch to the `FINISHED_RUNNING` state when we are certain that a workflow has finished running in Galaxy (not just relying on the History status of `OK`).

There may be an API call to ask Galaxy if a workflow has finished running. However, one other option I can think of that avoids this is to verify that all the output files defined by the workflow in IRIDA (e.g., available from the `getOutputs()` method of an `IridaWorkflowDescription` object https://github.com/phac-nml/irida/blob/6af8e5465830a4b78ac856a0fe2db4472190b996/src/main/java/ca/corefacility/bioinformatics/irida/model/workflow/description/IridaWorkflowDescription.java#L153

are available in Galaxy. That is, I'm thinking we can modify this code which checks if the IRIDA workflow/pipeline is complete https://github.com/phac-nml/irida/blob/6af8e5465830a4b78ac856a0fe2db4472190b996/src/main/java/ca/corefacility/bioinformatics/irida/service/impl/AnalysisExecutionScheduledTaskImpl.java#L274

to include an additional check where it asks Galaxy for a list of the History items and iterates over them, making sure that there exists datasets for the output files.

That way, a pipeline in IRIDA is completed only if it meets two conditions:

1. The Galaxy History status is reported as `OK`
2. A dataset entry in the Galaxy History exists for all output files.

I think this will avoid any possibility of IRIDA mistakenly thinking a pipeline has ended earlier than it really has.

Could you add the code to implement this solution (or if there's an easier solution, such as a Galaxy API call, please let me know).

**Comment by JeffreyThiessen** [comment]
>One comment I did have, I'm wondering if you could switch out runWorkflow with invokeWorkflow in these tests: AnalysisExecutionServiceGalaxyTest GalaxyWorkflowsIT

I fixed the ones in GalaxyWorkflowsIT, good catch. It looks like blend4j no longer returns those outputId's, so I think just removing the related test lines is all that's needed there.

The ones in AnalysisExecutionServiceGalaxyTest are not actually runWorkflow calls to blend4j, the irida object GalaxyWorkflowService has a runWorkflow method, and that's what's being tested there.

**Comment by JeffreyThiessen** [comment]
@apetkau Got that problem fixed. 

The easy way to test it is with a galaxy docker container.
Modify here after the try catch block https://github.com/galaxyproject/galaxy/blob/359fadc2e4366e9c85d5bf6af2573073aafddcbb/lib/galaxy/workflow/run.py#L173
with `import time; time.sleep(300)`

and you can reproduce the condition every time.

**Comment by apetkau** [comment]
@JeffreyThiessen could you merge `development` back into here and fix up the merge conflicts? Thanks.

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks so much for implementing this @JeffreyThiessen . Sorry about this taking so long to review.

I tested this on a Docker Galaxy (20.05) and the modification you suggested to reproduce the error, it works. I also tested it out on a Galaxy connected to our cluster (18.05) by submitted 200 analysis jobs at once, they all successfully ran with no issues. It looks like everything's working now.

There's only a few small comments below.

File: src/test/java/ca/corefacility/bioinformatics/irida/pipeline/upload/galaxy/integration/GalaxyWorkflowsIT.java [hunk 1]
```diff
@@ -350,15 +343,6 @@ public void testExecuteCollectionsPairedList() throws ExecutionManagerException
 		HistoryDetails historyDetails = historiesClient.showHistory(workflowOutput.getHistoryId());
 		assertNotNull(historyDetails);
 		
-		// outputs should exist
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply src/test/java/ca/corefacility/bioinformatics/irida/pipeline/upload/galaxy/integration/GalaxyWorkflowsIT.java:353]: Yes that is correct.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/services/scheduled/AnalysisScheduledTaskConfig.java [hunk 2]
```diff
@@ -46,6 +49,15 @@
 	@Autowired
 	private EmailController emailController;
 
+	@Autowired
+	private IridaWorkflowsService workflowsService;
+
+	@Autowired
+	private GalaxyHistoriesService galaxyHistoriesService;
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/services/scheduled/AnalysisScheduledTaskConfig.java]: Good catch, I think that was left over from while I was trying different solutions for this problem. Removed those lines now.

**Review by apetkau — APPROVED** [review]
This looks great to me. Thanks so much @JeffreyThiessen for all the time you spent on this and for the patience while waiting for me to review.

**Pull request merged**